### PR TITLE
add a feature flag for disabling external name service

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:375]: I think this may not work if a service moves to/from ExternalName, we may have stale data?

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:375]: Is that allowed? let me double check

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:375]: fixed

**Comment by GregHanson** [comment]
@hzxuzhonghu any reason for skipping the release note for this one?  I know you had concerns over the side-effects of including external name services ([link](https://github.com/istio/istio/pull/47566#issuecomment-1780298070)) - seems like a feature flag that others might be interested in come release time

**Comment by hzxuzhonghu** [comment]
OK, will add one

**Review by GregHanson — APPROVED** [review]

File: releasenotes/notes/47655.yaml
Comment (ericvn) [inline releasenotes/notes/47655.yaml:6]: Some minor changes, mostly to mention the default is true, and one may set it to false.
```suggestion
  **Added** `ENABLE_EXTERNAL_NAME_SERVICE` feature flag, which defaults to true conserving the current behavior processing the ExternalName service. It can be set to false removing the ExternalName service's influence completely.
```

**Review by howardjohn — COMMENTED** [review]
One thing we should be answering for any feature flag is "When will this be removed"? 

We are pushing to reduce the surface of feature flags, and how long they live. 

Its not clear this has a path to removal, making it likely a permanent shadow-API we are trying to reduce.

Its also not clear *why* we need this. What is the motivation for adding this? When should a user set this?

**Comment by hzxuzhonghu** [comment]
I donot think we should remove this as long as some users do not need istio to handle external name service

**Comment by howardjohn** [comment]
The question though is why disable this? There are 10,000 fields we support
in Istio, but no flag to turn each off.

I'm not necessarily against adding a flag, but we need to understand *why*
this needs to be skipped.

**Comment by costinm** [comment]
I agree that feature flag is not the right API for this. I also agree that for any behavior that is not identical to k8s we should have a proper config/API to enable ( and probably we should default to transparent k8s behavior - with backward compat in regular istio and correct behavior in ambient ). 

Too late for a lot of other things - but for this I would suggest MeshConfig.ProxyConfig. 

I would add that the different behaviors are not well documented - besides PR discussions - nor do we have consensus 
on what Istio should do.  So not even clear if this setting is a boolean or enum - and if global is the right granularity, but we can start with global.

 Why: a user should be able to install istio without any change in behavior.  If an istio feature changes something, it should be possible to turn it on/off or specify the behavior the user desire.

**Comment by hzxuzhonghu** [comment]
Not fully agree, firstly istio does not support all k8s features, at least some of those are not much required, and some are implemented by CNI like network policy.  
Secondly, external name service was badly supported in istio(not consistent with k8s), and recently changed with a lot of tricky transition, and also introduced incompatibility. For such feature, actually not many user requires, while at least we have serviceentry as an alternative.  Why not provide an option to disable it for 99% users while improving control plane performance

**Comment by costinm** [comment]
On Tue, Nov 7, 2023 at 12:17 AM Zhonghu Xu ***@***.***> wrote:

> Not fully agree, firstly istio does not support all k8s features, at least
> some of those are not much required, and some are implemented by CNI like
> network policy.
>
Depends on what you mean by "support" - but I hope we can agree that if an
app was previously working and installing Istio sidecar breaks it - it's a
bug
in Istio. We have quite a few - we may choose to call them "unsupported k8s
feature" - but in practice for a user it is a bug that will prevent the
adoption of Istio and break a particular application.

CNI and network policy are not 'supported' by Istio - nor is CSI and
persistent volumes, since Istio does not implement either of them - however
it would
be a bug if installing Istio would break an app using them. Istio sidecars
don't break CNI - ambient does, and while some may say 'ambient doesn't
support
network policy' to make it sound positive - reality is that any user of NP
would have a very bad surprise.



> Secondly, external name service was badly supported in istio(not
> consistent with k8s), and recently changed with a lot of tricky transition,
> and also introduced incompatibility. For such feature, actually not many
> user requires, while at least we have serviceentry as an alternative. Why
> not provide an option to disable it for 99% users while improving control
> plane performance
>
On this I agree - it was very buggy and broken before, and I think the new
behavior is also buggy ( by the definition above, i.e. an app not using
istio will change
after installing Istio ). ServiceEntry is an opt-in - like all other opt-in
APIs and features a user must explicitly configure things and will get
istio behavior.

I think an enum (in mesh config is possible) to select the behavior is a
good idea - with previously bad behavior as an option, current (perhaps
better) behavior
as another - and maybe we can add a 3rd option 'exactly what K8S does with
no change added by istio'.

For ambient there are expected changes in behavior for workloads opting in
to use ztunnel/waypoints - and it is the perfect place to default
to k8s behavior so users can turn on ambient by default in a cluster
knowing that nothing will break. I think most people agree with this, but
I believe even for Istio-sidecars it is very useful to have a mechanism to
select the behavior to match k8s and ambient - for migrations and
for new users who may want to use ambient for most workloads and sidecar
for select workloads ( but without differences in behavior ).

**Comment by hzxuzhonghu** [comment]
I completely agree with your comments above  @costinm

**Review by costinm — APPROVED** [review]
Considering the trade-offs discussed - I think this is not harming anyone but can help in some cases. We may refine it and have more options in the future - but env vars are experimental and supposed to change.

File: pilot/pkg/model/push_context.go
Comment (keithmattix) [inline pilot/pkg/model/push_context.go:1410]: This feels unintuitive to me; why would you need to set two flags in order to get sane external name functionality. If we eventually want ExternalNameAlias to become default eventually, what's the point of a separate flag?

**Review by keithmattix — COMMENTED** [review]
I'n generally ambivalent on this change. On one hand, having a way to say "Hey Istio, ignore this resource" is useful if there's cost or a negative effect associated with having Istio see it. On the other hand, it feels better to just fix that negative effect which I believe John's PR does. If I had to decide now, I'd err on the side of less options and not merge this unless there's a very compelling use-case

File: pilot/pkg/model/push_context.go
Comment (howardjohn) [reply pilot/pkg/model/push_context.go:1410]: EnableExternalNameAlias is already on by default which I think makes this PR less required

Comment (keithmattix) [reply pilot/pkg/model/push_context.go:1410]: Oh if that's the case then I'm a -1 on this PR. Something that's on by default shouldn't require another off-by-default flag to enable. And now that the first one has shipped, that would be a breaking change in 1.21

Comment (howardjohn) [reply pilot/pkg/model/push_context.go:1410]: This ENABLE_EXTERNAL_NAME_SERVICE is on by default so its not a breaking change.

However, there are literally 1000s of fields we read from in Kubernetes. I haven't seen a compelling argument for why *this* one needs a flag. I also think having flags of this nature in general (permanently one way without a deprecation path) are ideal.

Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go:1410]: EnableExternalNameAlias is to solve a lot of issues caused by external name service, but for those that do not need external name service at all.  Turn on `EnableExternalNameAlias ` is costly, it is the intention. 

I think it is agreed on a knob https://github.com/istio/istio/pull/46332#issuecomment-1780378016, and i have lots of concerns on that hack way.

**Pull request closed without merging**
